### PR TITLE
ORC-276: [C++] Create a simple tool to import CSV files

### PR DESCRIPTION
**Pull request by wgtmac**
A simple tool for users to create ORC files. It only support importing primitive type values from CSV files.

### CODE REVIEW

File: tools/src/CSVFileImport.cc [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+std::string extractColumn(std::string s, uint64_t colIndex) {
+  uint64_t col = 0;
+  size_t start = 0;
+  size_t end = s.find(DELIMITER);
+  while (col < colIndex && end != std::string::npos) {
+    start = end + 1;
+    end = s.find(DELIMITER, start);
+    ++col;
+  }
+  return s.substr(start, end - start);
```
Comment (xndai) [inline tools/src/CSVFileImport.cc]: You will subtract from string::npos when the last column doesn't end with a delimiter or the number of columns in csv is less than what specified in schema. We need to better handle these cases.

File: tools/src/CSVFileImport.cc [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+  orc::StringVectorBatch* stringBatch =
+    dynamic_cast<orc::StringVectorBatch*>(batch);
+  bool hasNull = false;
+  for (uint64_t i = 0; i < numValues; ++i) {
+    std::string col = extractColumn(data[i], colIndex);
+    if (col.empty()) {
+      stringBatch->notNull[i] = 0;
+      hasNull = true;
+    } else {
+      memcpy(buffer.data() + offset,
```
Comment (xndai) [inline tools/src/CSVFileImport.cc]: we might want to resize again when we are running out of 4M buffer size. Or at least we should raise an exception so we don't overflow the buffer.

**Comment by wgtmac** [comment]
Thanks @xndai. Please see the last commit for the fixes that you have mentioned.

@omalley @majetideepak Please take a look when you have time. Thanks!

**Comment by xndai** [comment]
LGTM. Thanks!

File: tools/src/CSVFileImport.cc [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+  ORC_UNIQUE_PTR<orc::Writer> writer =
+    orc::createWriter(*fileType, outStream.get(), options);
+  ORC_UNIQUE_PTR<orc::ColumnVectorBatch> rowBatch =
+    writer->createRowBatch(batchSize);
+
+  bool eof = false;
+  std::string line;
+  std::vector<std::string> data;
+  std::ifstream finput(input.c_str());
+  while (!eof) {
```
Comment (majetideepak) [inline tools/src/CSVFileImport.cc]: Some code comments will definitely help to extend this in the future.

File: tools/src/CSVFileImport.cc [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+  ORC_UNIQUE_PTR<orc::Type> fileType = ORC_NULLPTR;
+  if (schema.find(SCHEMA_PREFIX) != 0) {
+    std::cout << "Cannot find " << SCHEMA_PREFIX << " argument." << std::endl;
+    usage();
+    return 1;
+  } else {
+    fileType = orc::Type::buildTypeFromString(schema.substr(SCHEMA_PREFIX.size()));
+  }
+
+  if (argc > 4) {
```
[the remark this answers is not in the document]
Comment (wgtmac) [reply tools/src/CSVFileImport.cc]: Done! Good suggestion. Please review the new commit again. Thanks!

File: tools/src/CSVFileImport.cc [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+      tsBatch->data[i] = atoll(col.c_str());
+      tsBatch->nanoseconds[i] = 0;
+    }
+  }
+  tsBatch->hasNulls = hasNull;
+  tsBatch->numElements = numValues;
+}
+
+void usage() {
+  std::cout << "Usage: csv-import --input <input file> --output <output> "
```
[the remark this answers is not in the document]
Comment (wgtmac) [reply tools/src/CSVFileImport.cc]: Done. Thanks!

[the remark this answers is not in the document]
Comment (majetideepak) [reply tools/src/CSVFileImport.cc]: We need to include the short names too
```
std::cout << "Usage: csv-import [-h] [--help]\n"
            << "                [-d<character>] [--delimiter=<character>]\n"
            << "                [-s<size>] [--stripe=<size>]\n"
            << "                [-b<size>] [--block=<size>]\n"
            << "                [-b<size>] [--batch=<size>]\n"
            << "                <schema> <input> <output>\n"
            << "Import CSV file into an Orc file using the specified schema.\n"
            << "Compound types are not yet supported.\n";
```

**Comment by majetideepak** [comment]
I just realized we need to add tests for each tool as well.  `/tools/test` has some examples.
Some of the tools are missing tests as well. I will file a JIRA to cover those.
Sorry for not noticing this earlier.

**Comment by xndai** [comment]
Hi @majetideepak , Gang is on vacation and will look into your feedback after he's back next week. Thx.

**Comment by wgtmac** [comment]
@majetideepak Tests have been added, please review it when you get the chance. Thanks!

**Comment by majetideepak** [comment]
Looks good! Thank you!

**Pull request closed without merging**